llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,443 @@
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  bool Modified = renameMBBPhysRegs(LoopBBs);
+
+  return Modified;
+}
+
+bool AIEWawRegRewriter::renameMBBPhysRegs(
+    const SmallVector<const MachineBasicBlock *, 4> &LoopBBs) {
+  bool Modified = false;
+  for (const MachineBasicBlock *MBB : LoopBBs) {
+    LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+               dbgs() << "\n");
+    SmallVector<MCPhysReg, 16> UsedPhyRegs;
+    SmallVector<MCPhysReg, 16> UsedVirtRegs;
+    SmallVector<std::pair<MCRegister, MCRegister>, 8> ReplacementRegs;
+
+    SmallSet<MCPhysReg, 16> ForbiddenPhysRegs = getBBForbiddenReplacements(MBB);
+
+    std::map<unsigned, SmallVector<const MachineInstr *, 4>> UndefPartReuse;
+
+    for (const MachineInstr &MI : *MBB) {
+
+      if (isIdentityCopy(MI))
+        continue;
+
+      for (const MachineOperand &Op : MI.defs()) {
+        Register Reg = Op.getReg();
+        if (!Op.getReg().isVirtual())
+          continue;
+        if (VRM->hasRequiredPhys(Op.getReg()))
+          continue;
+        if (Op.isTied())
+          continue;
+        if (!TRI->isVecOrAccRegClass(MRI->getRegClass(Reg)))
+          continue;
+        //  Subregisters can reuse Constants in the high part of the
+        //  register, therefore renaming a reused physical register can
+        //  prohibit the elimination of a move of constants, thus degrading
+        //  performance if this physical register is removed
+        if (!DisableCopySubRegHandling &&
+            isUndefReserved(Op, MI, *MBB, UndefPartReuse))
+          continue;
+
+        if (isWAWCandidate(Reg, MI, UsedPhyRegs, UsedVirtRegs)) {
+          MCRegister PreSwapPhyReg = VRM->getPhys(Reg).id();
+          bool Replaced = replaceReg(Reg, ForbiddenPhysRegs);
+
+          if (Replaced) {
+            LLVM_DEBUG(dbgs() << MI);
+            Modified = true;
+            ReplacementRegs.push_back(
+                std::make_pair(PreSwapPhyReg, VRM->getPhys(Reg).id()));
+          }
+
+        } else {
+          // keep track of already visted physical and virtual Regs
+          UsedPhyRegs.push_back(VRM->getPhys(Reg));
+          UsedVirtRegs.push_back(Register::virtReg2Index(Reg));
+        }
+      }
+    }
+    // dump Reassignment
+    LLVM_DEBUG(dbgs() << "WAW Replacement Overview for bb." << MBB->getNumber();
+               dbgs() << "\n");
+    for (std::pair<MCRegister, MCRegister> &RegPair : ReplacementRegs) {
+      LLVM_DEBUG(dbgs() << printReg(RegPair.first, TRI, 0, MRI) << " -> "
+                        << printReg(RegPair.second, TRI, 0, MRI) << "\n");
+    }
+    LLVM_DEBUG(dbgs() << "\n");
+  }
+  return Modified;
+}
+
+bool AIEWawRegRewriter::isWAWCandidate(
+    const Register &Reg, const MachineInstr &MI,
+    const SmallVector<MCPhysReg, 16> &UsedPhyRegs,
+    const SmallVector<MCPhysReg, 16> &UsedVirtRegs) {
+  bool FoundPhyReg = UsedPhyRegs.end() != find(UsedPhyRegs, VRM->getPhys(Reg));
```

Comment:
nit: `const` in both `bool`'s

Reply:
add const to identify temp variables for reader's benefit